src/Sparrow.Server/Collections/SmallSet.cs

```diff
@@ -0,0 +1,281 @@
+        where TKey : unmanaged
+    {
+        private const int Invalid = -1;
+
+        private readonly int _length;
+        private readonly TKey[] _keys;
+        private readonly TValue[] _values;
+        private Dictionary<TKey, TValue> _overflowStorage;
+        private int _currentIdx;
+
+        public SmallSet(int size = 0)
+        {
+            _length = size > Vector<TKey>.Count ? size - size % Vector<TKey>.Count : Vector<TKey>.Count;
+            _keys = ArrayPool<TKey>.Shared.Rent(_length);
+            _values = ArrayPool<TValue>.Shared.Rent(_length);
+            _overflowStorage = null;
+            _currentIdx = -1;
+        }
+
+
+        public void Add(TKey key, TValue value)
+        {
+            int idx = FindKey(key);
+            if (idx == Invalid)
+                idx = RequestWritableBucket();
+
+            // We have overflowed already. 
+            if (_overflowStorage != null)
+                _overflowStorage[key] = value;
+
+            _keys[idx] = key;
+            _values[idx] = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindKey(TKey key)
+        {
+            int elementIdx = Math.Min(_currentIdx, _length - 1);
+
+            var keys = _keys;
+            if (Vector.IsHardwareAccelerated && elementIdx > Vector256<TKey>.Count)
+            {
+                Vector<TKey> chunk;
+                var keyVector = new Vector<TKey>(key);
+                while (elementIdx >= Vector256<TKey>.Count)
+                {
+                    // We subtract because we are going to use that even in the case when there are differences.
+                    elementIdx -= Vector256<TKey>.Count;
+                    chunk = new Vector<TKey>(keys, elementIdx + 1);
+                    chunk = Vector.Equals(keyVector, chunk);
+                    if (chunk == Vector<TKey>.Zero)
+                        continue;
+
+                    elementIdx = Math.Min(elementIdx + Vector256<TKey>.Count, _length - 1);
+                    goto Found;
+                }
+
+                chunk = new Vector<TKey>(keys);
+                chunk = Vector.Equals(keyVector, chunk);
+                if (chunk == Vector<TKey>.Zero)
+                    return Invalid;
+
+                elementIdx = Vector256<TKey>.Count - 1;
+
+            Found:
+                elementIdx = Math.Min(elementIdx, _currentIdx);
+            }
+
+            while (elementIdx >= 0)
+            {
+                ref TKey current = ref keys[elementIdx];
+                if (current.Equals(key))
+                    return elementIdx;
+
+                elementIdx--;
+            }
+
+            return Invalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int RequestWritableBucket()
+        {
+            if (_currentIdx >= _length && _overflowStorage == null)
+            {
+                var storage = new Dictionary<TKey, TValue>(_currentIdx * 2);
```

Comment:
Power of two instead of ` * 2` ?

Reply:
It could be, so you think it would change anything? This happens really really rarely, if at all. If this happens it is a bad sign (you are using a small set on a place that shouldnt be using this anyways).